front: adds warped map in simulation

jacomyal opened the pull request:
This PR solves issue #4882 (closes #4882).

Details:
- Adds `utils/useAsyncMemo` to help managing asynchronous states in React components
- Adds `src/common/Map/WarpedMap` that displays a map that is warped along a given path
  - `getWarping.ts` is a good start to understand how the warping works
  - `WarpedMap.tsx` is a reusable isolated component
  - `SimulationWarpedMap.tsx` is a component coupled to simulation state
- Isolates `TrainHoverPosition` to make it usable with a train that is not selected, and with a map that is not the "main" map
- Updates `SimulationResults` to mount a `SimulationWarpedMap` next to the `SpaceTimeChart`
- **Updates `SpaceTimeChart` so that it properly dispatches an `updateChart` event when the chart changes**

jacomyal committed:
simulation: drafts a warped map view left of GET

Details:
- Lots of new utils and helpers to get a transformation that "warps" a
part of the map to something straighter
- Hacky GeoJSON data loading using an hidden map
- Drafts a rendering of the final transformed data

jacomyal committed:
simulation: fixes missing points in warped map

jacomyal committed:
simulation: adds OSM data to warped map

jacomyal committed:
simulation: improves WarpedMap performances, adds some logs

jacomyal committed:
simulation: improves WarpedMap

Details:
- Deduplicates OSM queried features
- Improves rendering and data querying

jacomyal committed:
simulation: displays path and extremities in WarpedMap

jacomyal committed:
simulation: gradually improves displayed OSRD data

jacomyal committed:
simulation: refactors WarpedMap sources

Details:
- Simplifies code structure
- Adapts new imports rules (ie "no relative import")
- WarpedMap now renders already warped data, and no more calls any data
  or method from the store
- SimulationWarpedMap loads everything it needs from the store (current
  path, selected train...), handles warping the data, and gives
  everything to a WarpedMap

jacomyal committed:
simulation: highlights itinerary in WarpedMap

jacomyal committed:
simulation: displays trains in SimulationWarpedMap

Details:
- Adds utils/useAsyncMemo to simplify async data retrieving code
- Moves `getSimulationHoverPositions` and related helpers from
  SimulationResultsMap component code to new file
  SimulationResultsMap/helpers to make them reusable from elsewhere
- Uses those helpers to get similar trains to display in
  SimulationWarpedMap (after some warping magic)

jacomyal committed:
simulation: adds toggle button to warped map

jacomyal committed:
simulation: syncs warped map zoom and GET's zoom

Details:
- Updates SpaceTimeChart so that it calls `dispatchUpdateChart` when the
  graph is updated. I feel like it was missing, since this
  `dispatchUpdateChart` is called in `drawAllTrain` so the global chart
  is already "somehow" synchronized with the one in SpaceTimeChart
- Allows WarpedMap to be given a bounding box, in which case
  interactions are disabled
- Updates SimulationWarpedMap so that it computes a bounding box to fit
  SpaceTimeChart viewport, and gives it to WarpedMap

jacomyal committed:
simulation: fixes trains display on WarpedMap

Details:
- TrainHoverPosition no more reads train, allowancesSettings and
  viewport from global store, and requires them in props instead (this
  allows showing not the selected train, and not in the main map)
- Adds the proper trains in the TrainPosition that are not related to
  the selected train
- Adapts SimulationWarpedMap, WarpedMap and SimulationResultsMap to
  these changes

jacomyal commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+    const path = getAsyncMemoData(itineraryState);
+    if (!path || state.type !== 'dataLoaded') return [];
+
+    const transformedPath = state.transform(path) as typeof path;
```
This should be `warpedItinerary` instead (it's the same thing, so it's not bugged, it's just debt).

clarani commented on front/src/applications/operationalStudies/components/SimulationResults/SpaceTimeChart/SpaceTimeChart.jsx:
```diff
@@ -226,6 +226,8 @@ export default function SpaceTimeChart(props) {
       moveGridLinesOnMouseMove();
     }
+
+    dispatchUpdateChart(chart);
```
Can you add a comment so no one will remove it by accident ?
```suggestion
    // required to sync the warpedMap
    dispatchUpdateChart(chart);
```

clarani commented on front/src/common/Map/Layers/GeoJSONs.tsx:
```diff
@@ -386,38 +394,54 @@ const GeoJSONs: FC<{
+          ...source,
+          layers: source.layers.map((layer) => omit(layer, 'minzoom') as typeof layer),
+        }))
+      : res;
```
This test to keep or remove the minzoom should be in the function adaptFilter (maybe we should rename it) + could you add a small comment to precise why this test allows to render all the layers ? 
ex: remove the minzoom props so the layer always renders
```suggestion
    return layers
```

clarani commented on front/src/common/Map/Layers/OSM.tsx:
```diff
@@ -7,45 +7,49 @@ import mapStyleDarkJson from 'assets/mapstyles/OSMDarkStyle.json';
-      default:
-        return mapStyleJson as LayerProps[];
-    }
+export function getMapStyle(mapStyle: string): LayerProps[] {
```
This function is not used outside this file
```suggestion
function getMapStyle(mapStyle: string): LayerProps[] {
```

clarani commented on front/src/common/Map/WarpedMap/WarpedMap.tsx:
```diff
@@ -0,0 +1,200 @@
+/* eslint-disable no-console */
+import { useSelector } from 'react-redux';
+import React, { FC, PropsWithChildren, useEffect, useMemo, useState } from 'react';
```
:warning: linter warning
```suggestion
import React, { FC, useEffect, useMemo, useState } from 'react';
```

clarani commented on front/src/common/Map/Layers/OSM.tsx:
```diff
@@ -7,45 +7,49 @@ import mapStyleDarkJson from 'assets/mapstyles/OSMDarkStyle.json';
   }
+}
+
+export function genLayerProps(
```
```suggestion
export function genOsmLayerProps(
```

clarani commented on front/src/common/Map/Layers/OSM.tsx:
```diff
@@ -7,45 +7,49 @@ import mapStyleDarkJson from 'assets/mapstyles/OSMDarkStyle.json';
+  }));
+}
+
+export function genLayers(mapStyle: string, layerOrder?: number) {
```
```suggestion
export function genOsmLayers(mapStyle: string, layerOrder?: number) {
```

clarani commented on front/src/common/Map/WarpedMap/WarpedMap.tsx:
```diff
@@ -0,0 +1,200 @@
+}) => {
+  const prefix = 'warped/';
+  const [mapRef, setMapRef] = useState<MapRef | null>(null);
+  const { mapStyle, layersSettings, showIGNBDORTHO } = useSelector((s: RootState) => s.map);
```
Use selector
```suggestion
  const { mapStyle, layersSettings, showIGNBDORTHO } = useSelector(getMap);
```

clarani commented on front/src/common/Map/WarpedMap/WarpedMap.tsx:
```diff
@@ -0,0 +1,200 @@
+  switches: LAYER_GROUPS_ORDER[LAYERS.SWITCHES.GROUP],
+  track_sections: LAYER_GROUPS_ORDER[LAYERS.TRACKS_SCHEMATIC.GROUP],
+  // Unused:
+  catenaries: 0,
```
So if the layer order is 0, the layer is not loaded ?

jacomyal replied:
The `0`s are here for typing: `OSRD_LAYER_ORDERS` is a `Record<LayerType, number>` and requires all possible `LayerType`. If I use it as a `Partial<Record<LayerType, number>>`, I'll later have to deal with cases where `OSRD_LAYER_ORDERS` can miss the value.

clarani commented on front/src/common/Map/WarpedMap/WarpedMap.tsx:
```diff
@@ -0,0 +1,200 @@
+          ) as (OrderedLayerProps & {
+            'source-layer': string;
+          })[]
+        ).filter((layer) => !layer.id?.match(/-en$/)),
```
What do you do here ?
```suggestion
        ).filter((layer) => !layer.id?.match(/-en$/)),
```

jacomyal replied:
There are various OSM layers doubled with a `-en` suffixed clone. You can see this one for instance:

https://github.com/osrd-project/osrd/blob/7c7f180b11b475589d248a2df45fd44b4d09ed9a/front/src/assets/mapstyles/OSMStyle.json#L1457

I'll add a comment to clarify this.

clarani replied:
Thanks

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+  >({ type: 'idle' });
+  const pathfindingID = useSelector(
+    (s: RootState) => s.osrdsimulation.selectedProjection?.path
+  ) as number;
```
```suggestion
 const pathfindingID = useSelector(getSelectedProjection)?.path as number;
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+  const pathfindingID = useSelector(
+    (s: RootState) => s.osrdsimulation.selectedProjection?.path
+  ) as number;
+  const [getPath] = osrdEditoastApi.useLazyGetPathfindingByIdQuery();
```
It's better to use rtk definition like this, so we can more easily go to definition and see the types
```suggestion
  const [getPath] = osrdEditoastApi.endpoints.getPathfindingById.useLazyQuery();
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+    timePosition,
+    allowancesSettings,
+    simulation: { present: simulation },
+  } = useSelector((s: RootState) => s.osrdsimulation);
```
```suggestion
  } = useSelector(getSimulation);
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+    if (state.type !== 'dataLoaded') return null;
+
+    const foundTrain = simulation.trains.find((train) => train.id === selectedTrain.id);
+    if (!foundTrain) return null;
```
I think foundTrain is equal to selectedTrain, or am I missing smth ? If not, no need to recompute it, we can directly use foundTrain
```suggestion
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+  const selectedTrain = useSelector(getSelectedTrain);
+  const itineraryState: AsyncMemoState<Feature<LineString> | null> = useAsyncMemo(async () => {
+    if (!selectedTrain) return null;
+    if (state.type !== 'dataLoaded') return null;
```
```suggestion
    if (state.type !== 'dataLoaded' || !selectedTrain) return null;
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+                mapValues(osrdData, (collection: FeatureCollection) => state.transform(collection)),
+                isNil
+              ) as DataStatePayload['osrd'],
+            };
```
Suggestion : create a helper function like this:
```typescript
const transformDataStatePayload = (
  data: Partial<Record<string, FeatureCollection<Geometry, GeoJsonProperties>>>,
  transform: WarpingFunction
) =>
  omitBy(
    mapValues(data, (collection) => (collection ? transform(collection) : null)),
    isNil
  );
```

and then use it to make the code clearer
```suggestion
            const transformed = {
              osm: transformDataStatePayload(osmData, state.transform) as DataStatePayload['osm'],
              osrd: transformDataStatePayload(
                osrdData,
                state.transform
              ) as DataStatePayload['osrd'],
            };
```

clarani commented on front/src/common/Map/WarpedMap/getWarping.ts:
```diff
@@ -0,0 +1,58 @@
+export type WarpingFunction = ReturnType<typeof getWarping>['transform'];
+
+export default function getWarping(warpPath: Feature<LineString>) {
+  // Simplify the input path to get something "straighter", so that we can see
```
I'm not sure about this comment
```suggestion
  // Simplify the input path to get something "straighter", so that we can not see
```

jacomyal replied:
I can try to reformulate, but my version is good:
- If we warp along the exact given path, on the "straightened map", the input path will appear as a straight line.
- We simplify the path to warp along so that in the final "straightened map", the input path appears as still a little curvy (and it's intended)

So yes, we simplify the input path so that we can still see small curves of the initial path in the end.

clarani replied:
Ok got it, thanks !

clarani commented on front/src/common/Map/WarpedMap/core/helpers.ts:
```diff
@@ -0,0 +1,173 @@
+};
+export type PointsGrid = Record<number, Position>[];
+
+/**
+ * Path manipulation helpers:
+ */
```
Be carreful with the comments, with 2* you create documentation :)
```suggestion
/*
 * Path manipulation helpers:
 */

/** Given a line and a number N, cut the line into N samples (so compute N-1 points) and return the step between 2 points and the array of points */
```

clarani commented on front/src/common/Map/WarpedMap/core/helpers.ts:
```diff
@@ -0,0 +1,173 @@
+    } else if (i === samples - 1) {
+      points.push(point(last(line.geometry.coordinates) as Position));
+    } else {
+      const at = clamp(step * i, 0, l);
```
I think the clamping is not really required here ? (but you can keep it for corner cases I imagine)
```suggestion
      const at = step * i;
```

jacomyal replied:
Right, it seems useless indeed, I'll remove it.

clarani commented on front/src/common/Map/WarpedMap/core/helpers.ts:
```diff
@@ -0,0 +1,173 @@
+
+  return { step, points };
+}
+
+export function extendLine(line: Feature<LineString>, lengthToAdd: number): Feature<LineString> {
```
```suggestion

/** Given a line and a lengthToAdd, extend the line at its two extremities by lengthToAdd meters */
export function extendLine(line: Feature<LineString>, lengthToAdd: number): Feature<LineString> {
```

clarani commented on front/src/common/Map/WarpedMap/getWarping.ts:
```diff
@@ -0,0 +1,58 @@
+  // Extend the sample, so that we can warp things right before and right
+  // after the initial path:
+  const extendedSamples = extendLine(samplePath, sample.step);
+  const steps = extendedSamples.geometry.coordinates.length - 1;
```
stepCount/stepsCount is a better name I think
```suggestion
  const stepsCount = extendedSamples.geometry.coordinates.length - 1;
```

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+ * Using these two grids, it becomes possible to project any point from one grid
+ * to the other.
+ */
+export function getGrids(line: Feature<LineString>, params?: { stripsPerSide?: number }): Grids {
```
I think there is no need to over-engineer, maybe we can remove the param stripsPerSide and set a const StripesPerSide to 3, what do you think ? 
```suggestion
export function getGrids(line: Feature<LineString>, params?: { stripsPerSide?: number }): Grids {
```

jacomyal replied:
Agreed. Initially, all this warping code was in a standalone temp rep, where I tried to get everything tunable so that I end up with "good enough" parameters. Now, it seems overkill indeed.

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+ * to the other.
+ */
+export function getGrids(line: Feature<LineString>, params?: { stripsPerSide?: number }): Grids {
+  const l = line.geometry.coordinates.length;
```
```suggestion
  const pointsCount = line.geometry.coordinates.length;
```

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+  const flatGrid = featureCollection([]) as GridFeature;
+
+  // Generate flat line:
+  const flatLineStart = destination(c, totalLength / 2, 180);
```
Why 180 here ? I suppose it is 180 degree (so by default, bearing=0, it adds the point "above" the origin?), but anyway I think it lacks a comment :) 
```suggestion
  const flatLineStart = destination(c, totalLength / 2, 180);
```

jacomyal replied:
Here, `c` is the center of the new map, and `totalLength` will be its "height", in terms of path lengths.
The starting point of the line that will shape the grid we are looking for is in the bottom of the screen, and the distance between the two points will be `totalLength / 2`.
The bearing must be 180 to go "below", and 0 to go above (but above will be the ending point of the line).

I'll add a comment.

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+    flatLinePoints.push(destination(flatLineStart, -i * step, 180).geometry.coordinates);
+  const flatLine = lineString(flatLinePoints);
+
+  // Generate flat grid:
```
Maybe a schema here or in the description of the function would be useful
```suggestion
  // Generate flat grid:
```

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+    const p = line.geometry.coordinates[i];
+    const pP = line.geometry.coordinates[i === 0 ? i : i - 1];
+    const pN = line.geometry.coordinates[i === l - 1 ? i : i + 1];
+    const angle = bearing(pP, pN) + 90;
```
Can we discuss about this if you have some time ?
```suggestion
    const angle = bearing(pP, pN) + 90;
```

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+  // 2. Store triangles:
+  const grid = pointsGridToFeature(points);
+
+  return { warped: grid, regular: flatGrid };
```
The naming here is very confusing. I understand that grid is the "real" map and flatGrid the future warped one, correct ? I thought the warped map was the final one :confounded:
```suggestion
  return { warped: grid, regular: flatGrid };
```

jacomyal replied:
You are right. The main issue is that the map that is "straightened" is the one that is warped, while it feels to me that the less straight one (the original one) seems more twisted, so more warped...

I'll rename as "original" and "warped".

clarani replied:
Thanks !!

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+}
+
+/**
+ * This grid created by `getGrids` are a bit brute, and can have some weird
```
```suggestion
 * This grid created by `getGrids` is a bit brute, and can have some weird
```

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+
+/**
+ * Base helpers to manipulate grids:
+ */
```
```suggestion
/* Base helpers to manipulate grids: */

/** Given a grid (feature collection of triangles), return the dict of the triangles indexed by their id  */
```

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+  return keyBy(grid.features, (feature) => feature.properties.triangleId);
+}
+
+export function featureToPointsGrid(grid: GridFeature, steps: number): PointsGrid {
```
```suggestion
export function featureToPointsGrid(grid: GridFeature, stepCount: number): PointsGrid {
```

clarani replied:
```suggestion

/**
 * Given a grid (feature collection of triangles) and the sample number of the path,
 * return the matrix of points (array of dicts) from the grid
 */
export function featureToPointsGrid(grid: GridFeature, steps: number): PointsGrid {
```

clarani commented on front/src/common/Map/WarpedMap/core/quadtree.ts:
```diff
@@ -0,0 +1,112 @@
+      if (!quads.length) break;
+
+      const newQuads: QuadChild<Feature<G, P>>[] = [];
+      for (let j = 0, quadsCount = quads.length; j < quadsCount; j++) {
```
```suggestion
      for (let j = 0; j < quads.length; j++) {
```

jacomyal replied:
Here, your version does more lookups. It might seem like nothing, but in this `getQuadTree` function, it can be called thousands of times, and I tried to get the code as efficient as I could:
- All loops are vanilla `for (...;...;...)` loops (faster)
- I try to cache everything I can, including `quadsCount`

I think it's better, and I don't see the problem with it.

clarani replied:
OK, it was only a suggestion, I didn't know what was the reason behind this variable instantiation

clarani commented on front/src/common/Map/WarpedMap/core/grids.tsx:
```diff
@@ -0,0 +1,245 @@
+
+  return points;
+}
+export function pointsGridToFeature(points: PointsGrid): GridFeature {
```
No need to export this function (but you can add an empty line before it :) )
```suggestion

function pointsGridToFeature(points: PointsGrid): GridFeature {
```

clarani commented on front/src/common/Map/WarpedMap/core/quadtree.ts:
```diff
@@ -0,0 +1,112 @@
+  return !(mx1 > Mx2) && !(Mx1 < mx2) && !(my1 > My2) && !(My1 < my2);
+}
+
+export function getNewQuadChild<T>(box: BBox2d, isLeaf?: boolean): QuadChild<T> {
```
No need to export this function
```suggestion
function getNewQuadChild<T>(box: BBox2d, isLeaf?: boolean): QuadChild<T> {
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+      const latStart = (first(transformedPath.geometry.coordinates) as Position)[1];
+      const latEnd = (last(transformedPath.geometry.coordinates) as Position)[1];
+
+      /**
```
```suggestion
      /*
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+       * coordinate of the end of the path.
+       * Finally, `height` is the height in pixels of the SpaceTimeChart.
+       *
+       * Also, we now `latStart` and `latEnd`, which are the latitudes of the first and the last points of our
```
```suggestion
       * Also, we know `latStart` and `latEnd`, which are the latitudes of the first and the last points of our
```

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx:
```diff
@@ -0,0 +1,322 @@
+      allowancesSettings
+    ).map((position) => {
+      const transformedTrain = { ...position };
+      const pathLength = length(path);
+      const transformedPathLength = length(transformedPath);
```
These computations can be done outside the map, just before it
```suggestion
      const pathLength = length(path);
      const transformedPathLength = length(transformedPath);
```

jacomyal replied:
Good catch

clarani commented on front/src/common/Map/WarpedMap/core/helpers.ts:
```diff
@@ -0,0 +1,173 @@
+  return [a * x0 + b * x1 + c * x2, a * y0 + b * y1 + c * y2];
+}
+
+/**
```
```suggestion
/*
```

clarani commented on front/src/common/Map/WarpedMap/core/helpers.ts:
```diff
@@ -0,0 +1,173 @@
+ * Data helpers:
+ */
+const OSRD_BATCH_SIZE = 500;
+export async function getImprovedOSRDData(
```
If you have some time to explain me what this function does, it would be nice !

jacomyal replied:
Hahaha, I completely forgot it, sorry ^^
The explanations are here:

https://github.com/osrd-project/osrd/blob/2e4ed9e8172bf644462ea3062aff37dbd9016145/front/src/common/Map/WarpedMap/SimulationWarpedMap.tsx#L239-L243

Basically, before we call that function, OSRD data we have were loaded from a `DataLoader` map. This means that they probably are simplified, because they come from "vector servers" that simplify GeoJSONs before sending them, so that the payloads are way smaller.

That's why when you open a map first you can get so many objects showing fast, but when you zoom quickly, they are loaded again, with better precision.

But once I know what are the OSRD entities to show, I can load the "exact value" of each of them, so that when the user zooms deeply in the final map, they see the best possible version of these entities.

clarani replied:
Ok, maybe we should add a comment to explain what you wrote here, but I understand better now :)

clarani commented on front/src/common/Map/WarpedMap/SimulationWarpedMap.scss:
I think it would be nice to wrap all this css in ".simulationWarpedMap" class

clarani reviewed (changes requested):
Wow, nice job !! Thank you for this great work :muscle: 
I tested in local, 1 bug when you rotate the GET

jacomyal committed:
simulation: handles warped map review feedbacks

Most of the updates are:
- Adds various comments
- Clarifies getWarping output naming
- Fixes some comment blocks syntax

jacomyal committed:
simulation: fixed warped map when GET is rotated

clarani reviewed (approved):
LGTM !! Great job, tested in local :)
